File: game/ChainTiers.ts
// 分阶段强化：链长四档（2-3 / 4-5 / 6-7 / 8+），每档带一条数值倍率。
// 纯数据表：改表即生效，不改控件与执行器；倍率必须严格单调递增（由自检锁定）。

import { Config } from 'game/Config';

/** 一个链长档位的数据定义。 */
export interface ChainTierDef {
	id: string;
	/** 展示名（接触 / 连击 / 共鸣 / 超载）。 */
	name: string;
	/** 含下界。 */
	minChain: number;
	/** 含上界；最高档用大数表示“无上限”。 */
	maxChain: number;
	/** 数值倍率（严格单调递增）。 */
	multiplier: number;
}

/** 链长档位注册表：解析器只查表取倍率，不含任何按档位的分支。 */
export class ChainTiers {
	static readonly Contact = 'contact';
	static readonly Combo = 'combo';
	static readonly Resonance = 'resonance';
	static readonly Overload = 'overload';

	/** 最高档用该值表示“无上限”。 */
	static readonly NoLimit = 9999;

	/** 4 档：2-3 ×1.0 / 4-5 ×1.25 / 6-7 ×1.6 / 8+ ×2.1。 */
	static readonly List: ChainTierDef[] = [
		{ id: ChainTiers.Contact, name: '接触', minChain: Config.MinChainLength, maxChain: Config.MinChainLength + 1, multiplier: 1.0 },
		{ id: ChainTiers.Combo, name: '连击', minChain: 4, maxChain: 5, multiplier: 1.25 },
		{ id: ChainTiers.Resonance, name: '共鸣', minChain: 6, maxChain: 7, multiplier: 1.6 },
		{ id: ChainTiers.Overload, name: '超载', minChain: 8, maxChain: ChainTiers.NoLimit, multiplier: 2.1 },
	];

	static count(): number {
		return ChainTiers.List.length;
	}

	/** 取链长所在档位：低于首档下界回退首档，高于末档上界回退末档（永不返回 undefined）。 */
	static tierOf(chainLength: number): ChainTierDef {
		const list = ChainTiers.List;
		for (const tier of list) {
			if (chainLength >= tier.minChain && chainLength <= tier.maxChain) {
				return tier;
			}
		}
		if (chainLength < list[0].minChain) {
			return list[0];
		}
		return list[list.length - 1];
	}

	/** 链长对应的数值倍率（供解析器与自检使用）。 */
	static multiplierOf(chainLength: number): number {
		return ChainTiers.tierOf(chainLength).multiplier;
	}

	/** 链长对应档位的展示名。 */
	static nameOf(chainLength: number): string {
		return ChainTiers.tierOf(chainLength).name;
	}

	/** 倍率严格单调递增，且各档区间连续无空洞（自检锁定表格有效性）。 */
	static isStrictlyIncreasing(): boolean {
		const list = ChainTiers.List;
		if (list.length === 0) {
			return false;
		}
		for (let i = 0; i < list.length; i++) {
			if (list[i].maxChain < list[i].minChain) {
				return false;
			}
			if (i === 0) {
				continue;
			}
			if (list[i].multiplier <= list[i - 1].multiplier) {
				return false;
			}
			if (list[i].minChain !== list[i - 1].maxChain + 1) {
				return false;
			}
		}
		return true;
	}

	/** 全部档位的一行展示文本（供界面与自检报告）。 */
	static describeAll(): string {
		const parts: string[] = [];
		for (const tier of ChainTiers.List) {
			const upper = tier.maxChain >= ChainTiers.NoLimit ? '+' : '-' + tier.maxChain;
			parts.push(tier.minChain + upper + ' ' + tier.name + '×' + tier.multiplier);
		}
		return parts.join(' / ');
	}
}

File: game/ChainTiers.lua
-- [ts]: ChainTiers.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__Class = ____lualib.__TS__Class -- 1
local ____exports = {} -- 1
local ____Config = require("game.Config") -- 4
local Config = ____Config.Config -- 4
--- 链长档位注册表：解析器只查表取倍率，不含任何按档位的分支。
____exports.ChainTiers = __TS__Class() -- 20
local ChainTiers = ____exports.ChainTiers -- 20
ChainTiers.name = "ChainTiers" -- 20
function ChainTiers.prototype.____constructor(self) -- 20
end -- 20
function ChainTiers.count(self) -- 37
	return #____exports.ChainTiers.List -- 38
end -- 37
function ChainTiers.tierOf(self, chainLength) -- 42
	local list = ____exports.ChainTiers.List -- 43
	for ____, tier in ipairs(list) do -- 44
		if chainLength >= tier.minChain and chainLength <= tier.maxChain then -- 44
			return tier -- 46
		end -- 46
	end -- 46
	if chainLength < list[1].minChain then -- 46
		return list[1] -- 50
	end -- 50
	return list[#list] -- 52
end -- 42
function ChainTiers.multiplierOf(self, chainLength) -- 56
	return ____exports.ChainTiers:tierOf(chainLength).multiplier -- 57
end -- 56
function ChainTiers.nameOf(self, chainLength) -- 61
	return ____exports.ChainTiers:tierOf(chainLength).name -- 62
end -- 61
function ChainTiers.isStrictlyIncreasing(self) -- 66
	local list = ____exports.ChainTiers.List -- 67
	if #list == 0 then -- 67
		return false -- 69
	end -- 69
	do -- 69
		local i = 0 -- 71
		while i < #list do -- 71
			do -- 71
				if list[i + 1].maxChain < list[i + 1].minChain then -- 71
					return false -- 73
				end -- 73
				if i == 0 then -- 73
					goto __continue14 -- 76
				end -- 76
				if list[i + 1].multiplier <= list[i].multiplier then -- 76
					return false -- 79
				end -- 79
				if list[i + 1].minChain ~= list[i].maxChain + 1 then -- 79
					return false -- 82
				end -- 82
			end -- 82
			::__continue14:: -- 82
			i = i + 1 -- 71
		end -- 71
	end -- 71
	return true -- 85
end -- 66
function ChainTiers.describeAll(self) -- 89
	local parts = {} -- 90
	for ____, tier in ipairs(____exports.ChainTiers.List) do -- 91
		local upper = tier.maxChain >= ____exports.ChainTiers.NoLimit and "+" or "-" .. tostring(tier.maxChain) -- 92
		parts[#parts + 1] = ((((tostring(tier.minChain) .. upper) .. " ") .. tier.name) .. "×") .. tostring(tier.multiplier) -- 93
	end -- 93
	return table.concat(parts, " / ") -- 95
end -- 89
ChainTiers.Contact = "contact" -- 89
ChainTiers.Combo = "combo" -- 89
ChainTiers.Resonance = "resonance" -- 89
ChainTiers.Overload = "overload" -- 89
ChainTiers.NoLimit = 9999 -- 89
ChainTiers.List = {{ -- 89
	id = ____exports.ChainTiers.Contact, -- 31
	name = "接触", -- 31
	minChain = Config.MinChainLength, -- 31
	maxChain = Config.MinChainLength + 1, -- 31
	multiplier = 1 -- 31
}, { -- 31
	id = ____exports.ChainTiers.Combo, -- 32
	name = "连击", -- 32
	minChain = 4, -- 32
	maxChain = 5, -- 32
	multiplier = 1.25 -- 32
}, { -- 32
	id = ____exports.ChainTiers.Resonance, -- 33
	name = "共鸣", -- 33
	minChain = 6, -- 33
	maxChain = 7, -- 33
	multiplier = 1.6 -- 33
}, { -- 33
	id = ____exports.ChainTiers.Overload, -- 34
	name = "超载", -- 34
	minChain = 8, -- 34
	maxChain = ____exports.ChainTiers.NoLimit, -- 34
	multiplier = 2.1 -- 34
}} -- 34
return ____exports -- 34